Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/components/CaseDetailsController.java

```diff
@@ -78,6 +87,37 @@ public String getExportFilename() {
 		return parent.getContextData().getLogName() + ".csv";
 	}
 
+	/**
+	 * Update Activity <-> Attributes Map for quick lookup
+	 **/
+	private void updateActivityToAttributeMap(String caseId) {
+		activityToAttributeMap.clear();
+		APMLog log = ((LogDataWithAPMLog) parent.getLogData()).getFilteredAPMLog();
+		ATrace trace = log.get(caseId);
+		List<AActivity> activities = trace.getActivityList();
+		for (int i = 0; i < activities.size(); i++) {
+			AActivity activity = activities.get(i);
+			String activityName = activity.getName();
```

Comment:
move this to a service class. and return a JSONObject . and use toString oin this controller at the end.

Reply:
Because the trace visualization is done by ALog and ATrace, it'll be more consistent and safe to get this event attributes from ALog/ATrace rather than APMLog.